### PR TITLE
HTML-715: Add Program Attribute Tag 

### PR DESCRIPTION
**Pull request by ssmusoke**
https://issues.openmrs.org/browse/HTML-715

### CODE REVIEW

**Comment by ssmusoke** [comment]
@dkayiwa Please can you help me with this error for setting up the tests https://gist.github.com/ssmusoke/673dfde1a85283eb5a850831f95c369d

**Comment by dkayiwa** [comment]
@ssmusoke the error is caused by the fact that your `ProgramAttributeTagHandlerSupport1_10` implements `ConditionTagHandlerSupport` as a duplicate of what `ConditionTagHandlerSupport1_10` already implements in the same openmrs platform range `[1.10 - 2.1.*]`

**Comment by ssmusoke** [comment]
Thank you @dkayiwa that one is resolved now running into an issue with the OrderValidator 

`java.lang.IllegalStateException: Failed to load ApplicationContext

	at org.springframework.test.context.DefaultCacheAwareContextLoaderDelegate.loadContext(DefaultCacheAwareContextLoaderDelegate.java:94)
	at org.springframework.test.context.DefaultTestContext.getApplicationContext(DefaultTestContext.java:72)
	at org.springframework.test.context.support.DependencyInjectionTestExecutionListener.injectDependencies(DependencyInjectionTestExecutionListener.java:117)
	at org.springframework.test.context.support.DependencyInjectionTestExecutionListener.prepareTestInstance(DependencyInjectionTestExecutionListener.java:83)
	at org.springframework.test.context.TestContextManager.prepareTestInstance(TestContextManager.java:212)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.createTest(SpringJUnit4ClassRunner.java:200)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner$1.runReflectiveCall(SpringJUnit4ClassRunner.java:252)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.methodBlock(SpringJUnit4ClassRunner.java:254)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:217)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:83)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
	at org.springframework.test.context.junit4.statements.RunBeforeTestClassCallbacks.evaluate(RunBeforeTestClassCallbacks.java:61)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.springframework.test.context.junit4.statements.RunAfterTestClassCallbacks.evaluate(RunAfterTestClassCallbacks.java:68)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.run(SpringJUnit4ClassRunner.java:163)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:160)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:33)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:230)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:58)
Caused by: org.springframework.beans.factory.BeanDefinitionStoreException: Unexpected exception parsing XML document from class path resource [applicationContext-service.xml]; nested exception is org.springframework.context.annotation.ConflictingBeanDefinitionException: Annotation-specified bean name 'orderTypeValidator' for bean class [org.openmrs.validator.OrderTypeValidator] conflicts with existing, non-compatible bean definition of same name and class [org.openmrs.web.controller.order.OrderTypeValidator]
	at org.springframework.beans.factory.xml.XmlBeanDefinitionReader.doLoadBeanDefinitions(XmlBeanDefinitionReader.java:414)
	at org.springframework.beans.factory.xml.XmlBeanDefinitionReader.loadBeanDefinitions(XmlBeanDefinitionReader.java:336)
	at org.springframework.beans.factory.xml.XmlBeanDefinitionReader.loadBeanDefinitions(XmlBeanDefinitionReader.java:304)
	at org.springframework.beans.factory.support.AbstractBeanDefinitionReader.loadBeanDefinitions(AbstractBeanDefinitionReader.java:181)
	at org.springframework.beans.factory.support.AbstractBeanDefinitionReader.loadBeanDefinitions(AbstractBeanDefinitionReader.java:217)
	at org.springframework.beans.factory.support.AbstractBeanDefinitionReader.loadBeanDefinitions(AbstractBeanDefinitionReader.java:188)
	at org.springframework.test.context.support.GenericGroovyXmlContextLoader.loadBeanDefinitions(GenericGroovyXmlContextLoader.java:59)
	at org.springframework.test.context.support.AbstractGenericContextLoader.loadContext(AbstractGenericContextLoader.java:122)
	at org.springframework.test.context.support.AbstractGenericContextLoader.loadContext(AbstractGenericContextLoader.java:60)
	at org.springframework.test.context.support.AbstractDelegatingSmartContextLoader.delegateLoading(AbstractDelegatingSmartContextLoader.java:109)
	at org.springframework.test.context.support.AbstractDelegatingSmartContextLoader.loadContext(AbstractDelegatingSmartContextLoader.java:261)
	at org.springframework.test.context.DefaultCacheAwareContextLoaderDelegate.loadContextInternal(DefaultCacheAwareContextLoaderDelegate.java:68)
	at org.springframework.test.context.DefaultCacheAwareContextLoaderDelegate.loadContext(DefaultCacheAwareContextLoaderDelegate.java:86)
	... 25 more
Caused by: org.springframework.context.annotation.ConflictingBeanDefinitionException: Annotation-specified bean name 'orderTypeValidator' for bean class [org.openmrs.validator.OrderTypeValidator] conflicts with existing, non-compatible bean definition of same name and class [org.openmrs.web.controller.order.OrderTypeValidator]
	at org.springframework.context.annotation.ClassPathBeanDefinitionScanner.checkCandidate(ClassPathBeanDefinitionScanner.java:320)
	at org.springframework.context.annotation.ClassPathBeanDefinitionScanner.doScan(ClassPathBeanDefinitionScanner.java:259)
	at org.springframework.context.annotation.ComponentScanBeanDefinitionParser.parse(ComponentScanBeanDefinitionParser.java:87)
	at org.springframework.beans.factory.xml.NamespaceHandlerSupport.parse(NamespaceHandlerSupport.java:74)
	at org.springframework.beans.factory.xml.BeanDefinitionParserDelegate.parseCustomElement(BeanDefinitionParserDelegate.java:1427)
	at org.springframework.beans.factory.xml.BeanDefinitionParserDelegate.parseCustomElement(BeanDefinitionParserDelegate.java:1417)
	at org.springframework.beans.factory.xml.DefaultBeanDefinitionDocumentReader.parseBeanDefinitions(DefaultBeanDefinitionDocumentReader.java:174)
	at org.springframework.beans.factory.xml.DefaultBeanDefinitionDocumentReader.doRegisterBeanDefinitions(DefaultBeanDefinitionDocumentReader.java:144)
	at org.springframework.beans.factory.xml.DefaultBeanDefinitionDocumentReader.registerBeanDefinitions(DefaultBeanDefinitionDocumentReader.java:100)
	at org.springframework.beans.factory.xml.XmlBeanDefinitionReader.registerBeanDefinitions(XmlBeanDefinitionReader.java:510)
	at org.springframework.beans.factory.xml.XmlBeanDefinitionReader.doLoadBeanDefinitions(XmlBeanDefinitionReader.java:392)
	... 37 more
`

**Comment by mogoodrich** [comment]
Thanks @ssmusoke I'll take a look in the next couple of days at this PR!  If you haven't heard from me by next week feel free to ping me again.

**Comment by ssmusoke** [comment]
> Thanks @ssmusoke I'll take a look in the next couple of days at this PR! If you haven't heard from me by next week feel free to ping me again.

@mogoodrich Yes it will take me a couple of days to get all the issues sorted, I will definitely ping you for a review

**Comment by ssmusoke** [comment]
@dkayiwa Please can you help advise here too thanks in advance

**Comment by dkayiwa** [comment]
@ssmusoke add a `version` element to your `openmrs-web` dependency in the `api-2.2` project

**Comment by mogoodrich** [comment]
Okay, thanks @ssmusoke , just ping me when it's ready then, thanks!

**Comment by ssmusoke** [comment]
@dkayiwa I am back - I am now running into 

`Caused by: org.springframework.beans.factory.BeanDefinitionStoreException: Unexpected exception parsing XML document from class path resource [applicationContext-service.xml]; nested exception is org.springframework.context.annotation.ConflictingBeanDefinitionException: Annotation-specified bean name 'orderTypeValidator' for bean class [org.openmrs.validator.OrderTypeValidator] conflicts with existing, non-compatible bean definition of same name and class [org.openmrs.web.controller.order.OrderTypeValidator]
`

**Comment by dkayiwa** [comment]
That type of error cannot happen after adding the version to that dependency.

**Comment by ssmusoke** [comment]
@dkayiwa Looks like there is always a first time https://github.com/openmrs/openmrs-module-htmlformentry/blob/5b397402fcbf0f85086715b85e226b3530d040bb/api-2.2/pom.xml#L60

**Comment by dkayiwa** [comment]
Meaning?

**Comment by ssmusoke** [comment]
@dkayiwa I added a `version` element to my `openmrs-web` dependency in the `api-2.2` project but no effect still. I even tried the type `test-jar` and scope `tests` to no change

**Comment by dkayiwa** [comment]
The error message is different. Did you do more changes than just add the version?

**Comment by ssmusoke** [comment]
@dkayiwa I stripped out the unnecessary stack trace

**Comment by dkayiwa** [comment]
Have you looked at the travis log?

**Comment by ssmusoke** [comment]
@dkayiwa The travis errors are different from what I get on my computer, and they are not for anything that I have touched

**Comment by dkayiwa** [comment]
You could try doing `git reset --hard` and `git clean -df` before you compile again.

**Comment by ssmusoke** [comment]
@dkayiwa I have resolved the openmrs-web issues, however now the error that I see even on TravisCI is not connected at all to my test

**Comment by dkayiwa** [comment]
The compilation errors `cannot find symbol [ERROR]   symbol:   class HttpServletRequest` are caused by your removing the `openmrs-web` dependency.

**Comment by ssmusoke** [comment]
> The compilation errors `cannot find symbol [ERROR] symbol: class HttpServletRequest` are caused by your removing the `openmrs-web` dependency.

Thanks @dkayiwa That helped, let me keep yak shaving to get a proper error from my code

**Comment by ssmusoke** [comment]
@dkayiwa Happy Easter, wanted to pick your brain on the failures, since the failing tests are nothing that I have touched

**Comment by dkayiwa** [comment]
@ssmusoke the test failure error message `Cannot enroll in a program without specifying an Encounter Date or Enrollment Date` seems to come from the change that you made here: https://github.com/openmrs/openmrs-module-htmlformentry/blob/95d6c7aefe0cf4fa848d66b5c4e98bd0c5b8aa54/api/src/test/resources/org/openmrs/module/htmlformentry/include/enrollPatientInProgramForm.xml#L3

**Comment by ssmusoke** [comment]
> @ssmusoke the test failure error message `Cannot enroll in a program without specifying an Encounter Date or Enrollment Date` seems to come from the change that you made here:
> 
> https://github.com/openmrs/openmrs-module-htmlformentry/blob/95d6c7aefe0cf4fa848d66b5c4e98bd0c5b8aa54/api/src/test/resources/org/openmrs/module/htmlformentry/include/enrollPatientInProgramForm.xml#L3

Thanks @dkayiwa that sorts me out now let me work on the actual task at hand

**Comment by ssmusoke** [comment]
@dkayiwa Good morning I need your help again, I am getting a Class Not Found Exception below 

`Caused by: org.springframework.beans.factory.CannotLoadBeanClassException: Cannot find class [org.openmrs.layout.web.name.NameTemplate] for bean with name 'nameTemplateSpain' defined in URL [file:/Users/ssmusoke/IdeaProjects/openmrs-module-htmlformentry/api-tests/target/test-classes/TestingApplicationContext.xml]; nested exception is java.lang.ClassNotFoundException: org.openmrs.layout.web.name.NameTemplate`

yet I have included the api-tests sub-project which contains the relevant template and class

**Comment by dkayiwa** [comment]
@ssmusoke good morning to you too. Is this of help? https://issues.openmrs.org/browse/TRUNK-3849

**Comment by ssmusoke** [comment]
@dkayiwa That does not help since the api-tests have this reference https://github.com/openmrs/openmrs-module-htmlformentry/blob/788b4042411ad43a3e3334a989470214ff39a522/api-tests/src/test/resources/TestingApplicationContext.xml#L27

**Comment by dkayiwa** [comment]
If i wanted to reproduce that problem locally, how would i do it?

**Comment by ssmusoke** [comment]
@dkayiwa just pull this PR and run `mvn clean install`

**Comment by ssmusoke** [comment]
@dkayiwa Just checking in if you have any suggestions for me

**Comment by dkayiwa** [comment]
@ssmusoke remove the `-api-tests` dependency from the `api-2.2` pom.xml file.
Then add these dependencies https://pastebin.com/wWFYRzCq to the `api-2.2` pom.xml file

File: api-2.2/src/test/java/org/openmrs/htmlformentry/element/ProgramAttributeElementTest.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+	PatientService ps;
+
+	ProgramWorkflowService pws;
+
+	@Before
+	public void loadData() throws Exception {
+		ps = Context.getPatientService();
+		pws = Context.getProgramWorkflowService();
+		executeDataSet(XML_DATASET_PATH + new TestUtil().getTestDatasetFilename(PROGRAM_ATTRIBUTE_TEST_DATASET));
```
Comment (dkayiwa) [inline api-2.2/src/test/java/org/openmrs/htmlformentry/element/ProgramAttributeElementTest.java:35]: If after doing as i advised a couple of minutes ago, you get a missing `test-datasets.properties` file error, just change the above line to `executeDataSet(XML_DATASET_PATH + PROGRAM_ATTRIBUTE_TEST_DATASET)`

Comment (ssmusoke) [reply api-2.2/src/test/java/org/openmrs/htmlformentry/element/ProgramAttributeElementTest.java:35]: @dkayiwa Thank you very much, this has resolved the build failures, now to move to my tests

**Comment by ssmusoke** [comment]
@dkayiwa I am back, seems like now the tests cannot find the program_attribute_type table which should be available I think

**Comment by dkayiwa** [comment]
@ssmusoke `Failed tests:   testProgramAttributeWithProgramId(org.openmrs.htmlformentry.element.ProgramAttributeElementTest): expected:<0> but was:<2>`

**Comment by ssmusoke** [comment]
@dkayiwa strange I was getting that before now I am just getting 

`Tests in error: 
  testProgramAttributeWithProgramId(org.openmrs.htmlformentry.element.ProgramAttributeElementTest): org.dbunit.DatabaseUnitException: Exception processing table name='program_attribute_type'
`

**Comment by dkayiwa** [comment]
You could be having uncommitted changes. Travis agrees with me.

**Comment by ssmusoke** [comment]
@dkayiwa Please help me take a look at this https://github.com/openmrs/openmrs-module-htmlformentry/blob/48e7acc954f8c3d5171b7b44d6fdc97ad155e693/api-2.2/src/main/java/org/openmrs/module/htmlformentry/ProgramAttributeElement.java#L25-L70

I am adding the patientProgramAttribute to the PatientProgram to be updated, but it seems not be saving. 

When the tests are run - there is a single failure (in Travis too)

**Comment by ssmusoke** [comment]
@dkayiwa ping

**Comment by dkayiwa** [comment]
Can you share a link to the ticket?

**Comment by ssmusoke** [comment]
@dkayiwa done, I am only handling the first scenario where a patient is enrolled in a program, once I get this to work then I will move on to the other use cases, automatic enrollment etc

File: api-2.2/src/main/java/org/openmrs/module/htmlformentry/ProgramAttributeElement.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+public class ProgramAttributeElement implements HtmlGeneratorElement, FormSubmissionControllerAction {
+
+	private Program program;
+	private PatientProgram patientProgram;
+	private ProgramAttributeType programAttributeType;
+	private PatientProgramAttribute patientProgramAttribute;
+
+	public ProgramAttributeElement(FormEntryContext context, Map<String, String> parameters) {
+		try {
+			program = HtmlFormEntryUtil.getProgram(parameters.get("programId"));
```
Comment (dkayiwa) [inline api-2.2/src/main/java/org/openmrs/module/htmlformentry/ProgramAttributeElement.java:27]: In your tests, i did not see you pass the `programId` parameter anywhere.

Comment (ssmusoke) [reply api-2.2/src/main/java/org/openmrs/module/htmlformentry/ProgramAttributeElement.java:27]: @dkayiwa the file read is https://github.com/openmrs/openmrs-module-htmlformentry/blob/48e7acc954f8c3d5171b7b44d6fdc97ad155e693/api/src/test/resources/org/openmrs/module/htmlformentry/include/patientProgramAttributeEnrolledPatient.xml#L2 

Does that mean I have to pass the attributes too?

Comment (dkayiwa) [reply api-2.2/src/main/java/org/openmrs/module/htmlformentry/ProgramAttributeElement.java:27]: Shouldn't you also do a tag handler as in this commit? https://github.com/openmrs/openmrs-module-htmlformentry/commit/4339f81574c50ba86cbbca9055d7aeebd1980e10

**Comment by ssmusoke** [comment]
@dkayiwa I have updated to use the TagHandler approach but I am ending up with a NotYetPersistedException as per the stack trace https://gist.github.com/ssmusoke/3a37d2f34bbf5630750aa2b58777cd80

**Comment by ssmusoke** [comment]
@dkayiwa Ping for advice and help

**Comment by dkayiwa** [comment]
@ssmusoke did you see this pull request? https://github.com/openmrs/openmrs-module-htmlformentry/pull/167

**Comment by ssmusoke** [comment]
@dkayiwa so is this the preferred approach?

**Comment by dkayiwa** [comment]
@ssmusoke not really preferred. As long as it works.

**Comment by ssmusoke** [comment]
@dkayiwa That approach does not provide any assistance with the error that I am facing - can you help with that particular one

**Comment by dkayiwa** [comment]
@ssmusoke what i mean is simply that the pull request is implementing the same functionality that you are also trying to add. Just test it and give feedback, if any, to the pull request author. Does this make sense?

**Comment by ssmusoke** [comment]
@dkayiwa there are no steps to its usage and the approach is different since it's tied to the enroll into program tag

How would that affect the error that I am facing?

**Comment by dkayiwa** [comment]
@cliff are you implementing the same functionality as @ssmusoke? If yes, then we should have one pull request instead of two, dedicate our efforts on that.

**Comment by gitcliff** [comment]
@ssmusoke @dkayiwa the functionality that am implementing was to expand the existing `<enrollInProgram/>` to add program attributes to a patient already enrolled ...

**Comment by ssmusoke** [comment]
@dkayiwa any suggestions on how I can resolve the very specific issue that I have?

**Comment by dkayiwa** [comment]
@ssmusoke did you close this in order to go with the one of @gitcliff?

**Comment by ssmusoke** [comment]
Unfortunately not, I have abandoned this approach completely I am going to hack it differently

**Pull request closed without merging**